### PR TITLE
fix(e2e): bound descendants that setsid out of every signalable group (#861)

### LINKED ISSUE #861: e2e harness: DOT_AGENT_DECK_TEST_MAX_LIFETIME_SECS is not enforced once the owning test process dies
## What was observed

On the shared remote dev box (`dot-agent`):

```
pid      : 2043710
cmd      : /bin/bash -c source /var/tmp/dad-e2e-1000/dad-tests-2042039-Zn1Xcx/.tmpuyAZ3I/home/.claude/shell-...
elapsed  : 4 days 01:23
PPID     : 1                                     (reparented to init)
env      : DOT_AGENT_DECK_TEST_MAX_LIFETIME_SECS=300
children : 1
```

- The cap in the process's own environment is **300 s**. It had been alive roughly 350,000 s — about **1170x** the cap.
- Its owning test process (pid `2042039`, per the `dad-tests-<pid>-*` root name) was **dead**.
- Its temp root `/var/tmp/dad-e2e-1000/dad-tests-2042039-Zn1Xcx` had already been removed.

## Why this looks wrong

The cap is present in the child's environment but appears to be enforced by a watchdog living in the owning test process. When the owner dies first, the child reparents to init and nothing enforces the deadline.

## Knock-on effect on CLAUDE.md rule 14

Rule 14 derives `clean-e2e-tmp`'s 30-minute dead-PID floor from this cap:

> the deck's lazily-spawned daemon `setsid`s out of that process group and can outlive it by up to `DOT_AGENT_DECK_TEST_MAX_LIFETIME_SECS` while still writing under the root

"up to the cap" is the assumption the floor rests on, and this process outlived it by four days. That does not by itself make the floor unsafe — the temp root here was already deleted, so there was nothing left to reap — but the reasoning behind `DEAD_PID_MIN_AGE` should be re-checked, or re-derived from something that actually holds.

## Suggested fix direction

Enforcement that survives the owner's death rather than a watchdog in the parent — the child arming its own deadline, or a `prctl(PR_SET_PDEATHSIG)` / process-group mechanism.

## Test obligation

Per CLAUDE.md rule 5, if the fix lands in `tests/common/mod.rs` that file is linked by all 71 `tests/e2e_*.rs` files, so this needs the broad local pass rather than a scoped one. Worth accounting for when scheduling it — see the ordering note in the companion issues.

Found while diagnosing unrelated slowness on that box.

### PR DESCRIPTION
**Pull request by vfarcic**
Closes #861.

## The defect, measured

`DOT_AGENT_DECK_TEST_MAX_LIFETIME_SECS` is read in exactly three places — `daemon.rs`'s backstop task, `wrap::arm_wrap_self_defense`, and `wrap::arm_child_group_backstop`'s forked reaper. Each bounds the process it runs in and reaches everything *else* by signalling a **process group**.

A `killpg` reaches nothing that has left the group, and agents put processes outside it as a matter of routine. Measured live on the dev box, from inside Claude Code's own Bash tool:

```
pid=186393 ppid=184221 pgid=186393 sid=186393 comm=bash     <- own session
pid=184221 ppid=184166 pgid=183823 sid=183823 comm=claude    <- the agent's session
```

Every shell Claude Code runs leads a session of its own — the behaviour `agent_pty::close_agent` already names as "the `setsid`'d sub-shells Claude Code creates internally". Such a process inherits the cap and was bounded by nothing. The reported orphan: pid 2043710, `DOT_AGENT_DECK_TEST_MAX_LIFETIME_SECS=300` in its own environment, `PPID 1`, alive **4 days 01:23** — about 1170x its cap — with its owning test process dead and its temp root already removed from under it.

**`wrap` could not have covered it, and that is the half worth noticing.** Only `IntegrationStrategy::Wrapper` agents are wrapped. Codex is; Claude Code is `NativeHooks` and is spawned bare, so that pane had no wrapper, no `arm_wrap_self_defense` and no forked reaper — only the daemon's registry `killpg`, which the escapee is outside of by construction.

## Reproduction first

`tests/agent_lifetime_bound.rs::a_setsid_escapee_is_still_bounded_by_the_cap` builds the shape with `setsid(1)` standing in for the agent, SIGKILLs the wrapper so no reap loop runs, and asserts the escapee is gone inside the budget. Against `main` it failed for the reported reason:

```
assertion `left == right` failed: a SIGKILL'd wrapper stranded a descendant that had
`setsid`'d out of the wrapped child's process group, with the lifetime cap armed.
  left: Some(true)
 right: Some(false)
```

Its control — the same probe with no cap in the environment — asserts the escapee **survives**, so the assertion above cannot pass for a reason other than the cap.

## The mechanism

Both spawn seams export a per-spawn `DOT_AGENT_DECK_TEST_LIFETIME_TAG` beside the cap, and the reaper sweeps `/proc/<pid>/environ` for it at the deadline in addition to `killpg`ing the group. A tag survives what a group cannot: `execve` passes the caller's environment on by default, so it reaches descendants of every shape unless one deliberately rebuilds its own, and it is still readable after the whole ancestry is gone — which is the state the reported orphan was found in. A parent-chain walk cannot substitute: by the time the deadline expires the tree has already come apart.

Two consequences worth reading in the diff rather than skipping:

- **Group death no longer ends the reaper.** It used to `_exit(0)` the moment the child's group was gone, on the reasoning that nothing was left to bound. An escapee falsifies that. Group death now triggers one sweep — the reaper still exits immediately if nothing carries the tag, and otherwise stays to its deadline and ends what does, because the escapee's own cap says it may run until then.
- **Armed at `agent_pty::spawn`, not only in `wrap`.** That is the seam every launch path funnels through, so one call covers wrapped and unwrapped agents, the daemon's panes, the TUI's, and a test's in-process registry alike. It is applied straight to the `CommandBuilder` rather than through `opts.env`, so `spawn_env` does not capture it and a respawn mints its own rather than replaying a tag whose reaper died with the previous generation.

Sweep safety rests on one property: inheritance is the only route a tag travels, and every minted value is unique (pid + per-process counter + `CLOCK_MONOTONIC` nanos, unit-tested over 1000 mints). So a sweep reaches only processes that inherited the tag from the one child it was armed for — not a sibling pane, not a concurrent run, not a developer's own agents.

## How it degrades

- **Linux-only.** Reading another process's environment needs `/proc/<pid>/environ`; `signal_tagged` is a no-op returning `0` elsewhere and the pre-existing group half is unchanged. Same platform split `wrap`'s reaper already has between `close_range(2)` and a portable `close` loop. Nothing new is `#[cfg(unix)]`-only that was not already.
- **A `SIGKILL` aimed at the reaper itself leaves the deadline with no holder.** That residual is unchanged and pre-existing: the reaper leaves its parent's process group so a group kill cannot take it, but a direct signal still can.
- **A descendant that rebuilds its environment from scratch drops the tag** and is not found.
- **Gated on the variable, not the build.** No tag is minted, exported or swept unless the cap parses — the harness sets it and production does not. Note the qualifier is on the *variable*, exactly as it already is for the cap: a developer who exports it and starts an ordinary deck gets this machinery too. Measured on the built binary, with a positive control so the negative result means something:

```
PRODUCTION (cap unset): tag_in_child_environ=0  reapers_added=0
TEST       (cap=15)   : tag_in_child_environ=1  reapers_added=1
```

## Rule 14: the `DEAD_PID_MIN_AGE` derivation was re-checked, and the numbers are unchanged

The floor's stated derivation was *"an orphan can keep writing under the dead test's root for up to that cap"*. That premise holds for the class it names — the daemon, and a wrapped agent's child since #657/#661 — and #861 falsified it for an escapee.

It did not make the floor unsafe, for two compounding reasons. No finite floor bounds an unbounded writer, so a larger number would have bought nothing. And the floor never protected that class: `clean-e2e-tmp` reads `live-pid` off the **test process's** PID in the root's name and never off an orphan, so an escapee has never held its own root against `--apply` at any age. That was already measured and accepted — #668 censused 221 such orphans, *"each still holding a working directory the tooling had already deleted"*, naming the cost as the process rather than the deletion.

So `MAX_PINNED_ORPHAN_CAP_SECS` and `DEAD_PID_MIN_AGE` are **untouched** and linkage-check rule 11 is untouched (verified green). What changed is the claim, narrowed in the three places that stated it: the constant's own doc, `docs/develop/e2e-temp-dirs.md`, and CLAUDE.md rule 14. The fix then leaves the derivation *better* supported than when it was written — an escapee carrying the tag is now bounded by the same cap on Linux while its reaper lives — and still not universal, which the prose now says.

No recycled-PID branch was added, in any form. Rule 14's deviation stands.

## Rule 12: did this change the TUI↔daemon contract?

**No.** No wire shape moved, no field's meaning shifted, no handler behaviour changed. `PROTOCOL_VERSION` is unchanged and no `.breaking.md` fragment applies. The added code paths are gated on a variable production never sets, so an old daemon and a new TUI interoperate exactly as before — measured, per the bullet above.

**Rule 12's cross-version manual test HAS now been run, and it passes.** This paragraph previously recorded an abstention; that is superseded. The old side is the real **v0.39.2** release binary (an ancestor of this branch, so the skew is unidirectional) rather than a build-id override, since an override runs the same code on both ends and cannot observe a semantic break behind a stable wire. Sequence as prescribed: v0.39.2 daemon → v0.39.2 TUI brings the orchestration roles up → that TUI closed, daemon and agents left running → branch TUI → mismatch prompt naming both live agents → declined. Confirmed the scenario was genuinely reached rather than silently downgraded to a same-version run: **exactly one** `Attach protocol listening` line in the isolated log, one `daemon serve` process, and the same daemon pid with `/proc/<pid>/exe` pointing at the v0.39.2 binary before and after the branch TUI attached (`running daemon: 0.39.2-g364a182` vs `this binary: 0.39.2-g0e40f70f`). Results: **a delegate still routes** (rc 0; the old daemon logged `Received delegate signal`, wrote the task pointer into the worker pane's PTY and the sentinel-bearing task text to `worker-task-coder.md`), and **hooks still arrive** — `work-done` rc 0 with the old daemon logging `retired the outstanding delegation`, so it was matched to the commission the delegate armed and the full round trip closed across the version boundary, with the orchestrator pane receiving `Worker coder has completed their task.`; the status hook (`agent-event --type running`) was run **last** per the rule's own note, rc 0; both agents still registered afterwards. All five isolation variables the rule names were pinned, `DOT_AGENT_DECK_LOG` and `DOT_AGENT_DECK_EXPERIMENTAL` included. **No contract break**, so `PROTOCOL_VERSION` unchanged is correct and no `.breaking.md` applies. Limits stated rather than glossed: the roles are deterministic stand-in scripts rather than real agents (this gate is about the TUI↔daemon contract, so delivery was made observable on disk), one previous release, Linux only. Full evidence is in the review thread on `src/agent_pty.rs`.

## Tests run

Fast tier and the tests covering the change, named:

| what | result |
| --- | --- |
| `cargo fmt --check` | pass |
| `cargo clippy --workspace --all-targets --features e2e,e2e-live -- -D warnings` | **pass, exit 0** — all four flags, on this exact tree |
| `cargo xtask linkage-check` | pass, 11 rules — incl. rule 11's cap/floor pairing, re-run on the final tree |
| the 15 fast-tier binaries that spawn agents (below) | **161 tests, 161 passed, 0 skipped** |
| `cargo test-fast` in full | **not completed locally; green in CI on all three platforms** — see the note under the table |

The covering set, named, since "ran the related tests" is not a claim anyone can check — every fast-tier file that constructs an `AgentPtyRegistry`, calls `run_daemon_with`, or drives `wrap`:

`agent_lifetime_bound`, `wrap_io`, `shell_activity`, `rehydration`, `daemon_protocol`, `daemon_status`, `delegate_respawn_recovery`, `delegate_prompt_injection`, `orchestration_delegate`, `orchestration_orphan`, `idle_worker_detector`, `work_done_reporting`, `foreign_pane_detection`, `agent_detection`, `spawn_time_role_prompt_submit_after_session_start`.

That set is chosen for coverage, not convenience. `wrap_io` holds the pre-existing backstop tests (`wrap_child_group_backstop_reaps_a_child_stranded_by_a_sigkilled_wrapper`, `wrap_max_lifetime_backstop_ends_an_unsignalled_wrapper_and_its_child`) — the behaviour this PR modifies. `shell_activity` holds the real `setsid`'d-grandchild tests. And `daemon_status`, `delegate_respawn_recovery` and `delegate_prompt_injection` reach `spawn_inprocess_daemon`, so they exercise the new `fork` from inside a live multi-threaded Tokio runtime, which is the sharpest risk in the diff.

Individually, the new and pre-existing lifetime tests: `a_setsid_escapee_is_still_bounded_by_the_cap` (new — red against `main`), `an_unwrapped_agent_spawn_carries_a_lifetime_tag` (new), `a_term_resistant_wrapped_child_is_still_bounded_by_the_cap`, `in_process_registry_spawn_arms_the_wrapped_child_lifetime_bound`, `ambient_lifetime_caps_are_clamped_to_the_reapers_ceiling`, plus `lifetime_tag::tests::*` (7 unit tests, one of which spawns a real tagged child and an untagged control and sweeps with signal `0`).

**`cargo test-fast` in full did not complete on my machine, and I am not going to pretend otherwise.** It was killed four times for low system memory — always during the build, never during a test — because a sibling agent was running a large `--features e2e` workspace build on the same 28 GB box; free memory repeatedly fell under 600 MB with swap at 5 GB of 8, and lowering `CARGO_BUILD_JOBS` to 6 and then 3 did not save it. What I ran locally instead is the scoped set above, which is rule 5's second tier and the part that actually covers this diff.

**CI then ran the tier in full, and it is green on all three platforms** — which is the observation of record, so read it rather than my word. From the Linux `build` job on the final commit: `cargo nextest run --workspace` → `Starting 2411 tests across 139 binaries` → `Summary [113.812s] 2411 tests run: 2411 passed, 0 skipped`. `build-macos` and `build-windows` run the same command per rule 5 and both pass, so the tier is covered on the two platforms no local gate can reach at all. All four required checks (`build`, `build-macos`, `build-windows`, `security`) are green.

**Each half was proved load-bearing by reverting it and watching the test go red**, one at a time:

| reverted | what went red |
| --- | --- |
| the `agent_pty::spawn` tag export | `an_unwrapped_agent_spawn_carries_a_lifetime_tag` — the stand-in recorded `"<unset>"` |
| the group-death latch (restore the old unconditional `_exit(0)`) | `a_setsid_escapee_is_still_bounded_by_the_cap`, `ExitAtOnce` case |
| the deadline sweep itself | `a_setsid_escapee_is_still_bounded_by_the_cap`, **both** cases: `["wrapped child exited at once", "wrapped child kept running"]` |

`a_term_resistant_wrapped_child_is_still_bounded_by_the_cap` stayed green through all three, so no revert was breaking the group half by accident.

The escapee test runs both fates before asserting rather than in a loop, deliberately: they fail for different reasons, and a loop that panicked on the first would leave the second's assertion never observed failing — which the third row above is the evidence against.

`e2e-deterministic` is red on `main` and on every branch from #818, and is not a required check — any red there is that, not this.

## Review rounds

**Round 1 — `build-macos`, and it is the divergence rule 2 documents.**

`build-macos` went red on the first push at `cargo clippy -- -D warnings`: `clippy::missing_safety_doc` on the `#[cfg(all(unix, not(target_os = "linux")))]` arm of `signal_tagged`, a `pub unsafe fn` whose Linux twin carries a `# Safety` section and whose stub did not. My local four-flag run was clean precisely because it never lints that configuration — rule 2 says so in as many words. Fixed in `ecb5ce1` with the stub's contract spelled out (it signals nothing and reads nothing; `unsafe` stays only so the `fork`ed reaper is not handed a differently-shaped function per target). `windows-cross-check` had already passed, so the Windows configuration was never affected.

**Round 2 — Greptile, four findings, three of them right.** All four are answered inline; the substance:

- **P1, overbroad tag match — valid, and a real defect in code that sends `SIGKILL`.** The needle was a bare `NAME=VALUE`, so a tag whose value is a *prefix* of another tag's matched that other tag's entry, as did the tag's text sitting inside an unrelated variable's value. It is now the exact NUL-delimited entry `\0NAME=VALUE\0`, over-long values are refused rather than truncated (a truncated needle *is* a prefix), and there are two regression tests — one on the shapes directly, one spawning a live decoy whose value strictly extends ours. This shipped in my first commit because I reasoned about the tag's *uniqueness* and never about its *boundaries*.
- **P2, unproven uniqueness claim — valid, and my own rule 17 violation.** Narrowed to "no two tags held by *live* processes collide", with the cross-boot collision named and shown inert. Same narrowing in the changelog.
- **P1, cleanup can kill strangers — partly acted on.** The *group* send is gone; cleanup is a single-pid `SIGKILL`. I declined the suggested pidfd: this repository has already reasoned about that residual and deliberately bounded rather than closed it (`wrap.rs`'s `CHILD_GROUP_BACKSTOP_POLL` doc), and the sibling probe in the same file carries the identical residual — one file with two answers to one question is worse than the residual.
- **P2, cross-version test — not fixed, deferred to a maintainer, thread left open.** My reasoning and the measurement are in the thread and in the rule 12 section above. Rule 12's own text warns that a mis-sequenced run yields a false green worse than no run, and the correct sequence needs two TUIs driven interactively, which I cannot do reliably here. This is a judgment call about the rule, not about the diff.

Note Greptile has `triggerOnUpdates: false`, so it will not re-review the push that fixed these — which is why each thread has a written reply rather than a second pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lifetime_tag.rs [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+/// window, so the worst case is a few hundred thousand byte compares per
+/// process, paid at most twice per reaper.
+#[cfg(target_os = "linux")]
+fn contains(haystack: &[u8], needle: &[u8]) -> bool {
+    if needle.is_empty() || haystack.len() < needle.len() {
+        return false;
+    }
+    haystack
+        .windows(needle.len())
+        .any(|window| window == needle)
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply src/lifetime_tag.rs:561]: Valid, fixed. You are right that this matters more here than an overbroad match usually would, because a match is followed by SIGTERM and then SIGKILL.

The needle is now the whole entry with both boundaries — `\0NAME=VALUE\0` rather than `NAME=VALUE` — built in one place (`LifetimeTag::from_value`, shared by `mint` and `inherited`). Since the first entry in an environment block has no NUL in front of it, `environ_contains` primes its streaming window with one synthetic NUL, so every entry is uniformly preceded by a boundary at no cost.

Two things I made explicit rather than leaving implicit:

- an over-long value is now **refused** rather than truncated, because a truncated needle *is* a prefix — the exact shape the boundaries exist to reject;
- `inherited()` also refuses an empty value (which would build `\0NAME=\0` and match any process carrying the variable empty) and one containing a NUL.

Regression tests, both new: `a_tag_that_is_a_prefix_of_another_does_not_match_it` covers the prefix case, the reverse case, and your second shape — the tag text buried inside an unrelated variable's value. And the real-process sweep test now spawns a third child whose tag value strictly *extends* ours and asserts the sweep still matches exactly one, so the property is checked against a live `/proc` and not only a fixture.

One residual I would rather state than hide: a process that has rewritten its own environ into a block with no trailing NUL will not match its last entry. That is a benign false negative — it declines to signal rather than signalling the wrong thing — and it is documented at `environ_contains`.

File: tests/agent_lifetime_bound.rs [hunk 2]
```diff
@@ -364,3 +448,231 @@ fn a_term_resistant_wrapped_child_is_still_bounded_by_the_cap() {
...
+    // regression in the code under test would itself mint the orphan #861 is
+    // about.
+    if common::process_running(escapee_pid) {
+        // SAFETY: best-effort cleanup of a pid this test created. Same
+        // check-then-act residual as every other site here: between
+        // `process_running` and the signal the pid could be reaped and
+        // reissued, so this is bounded same-UID exposure, not an impossibility.
+        unsafe {
+            libc::kill(-escapee_pid, libc::SIGKILL);
+            libc::kill(escapee_pid, libc::SIGKILL);
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply tests/agent_lifetime_bound.rs:604]: Partly acted on, and I want to be precise about which part.

You are right that the *group* send is the dangerous half, and it is gone. Cleanup is now a single `kill(escapee_pid, SIGKILL)` rather than `kill(-pid)` followed by `kill(pid)`, so a recycled pid can at worst cost one stranger instead of a whole group of them. That narrowing is free here: this escapee's only children are the transient `sleep 1`s of its own loop, which end within a second of it whether they are signalled or not.

I did **not** move to a pidfd, and the reason is consistency rather than convenience. The check-then-act residual on a bare pid is one this repository has already reasoned about at length and decided to bound rather than close — `wrap.rs`'s `CHILD_GROUP_BACKSTOP_POLL` doc states it as "a bounded same-UID residual, not an impossibility" and concludes that a strict guarantee "would need an OS-owned containment mechanism that names the processes rather than a number — a dedicated cgroup or equivalent", which is explicitly out of scope for that backstop. The sibling probe in this same file carries the identical residual with the identical comment. Introducing a pidfd in one test's cleanup path would leave two different answers to the same question in one file, and the one place it actually matters is the production reaper, not a test's best-effort teardown.

What I have added instead is the reasoning at the call site, so the next reader sees why the narrow form was chosen and what is still open, rather than inferring that nobody thought about it. If you would rather see a pidfd applied uniformly — the reaper included — that is a real change worth its own issue, and I would not want to make it half-way in a test.

File: src/lifetime_tag.rs [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+        use std::sync::atomic::{AtomicU64, Ordering};
+        static SEQ: AtomicU64 = AtomicU64::new(0);
+
+        let seq = SEQ.fetch_add(1, Ordering::Relaxed);
+        let text = format!(
+            "{DOT_AGENT_DECK_TEST_LIFETIME_TAG}={:x}-{:x}-{:x}",
+            std::process::id(),
+            seq,
+            monotonic_nanos()
+        );
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply src/lifetime_tag.rs]: Valid, fixed, and the rule-17 citation is fair — I wrote an absolute about an implementation and it was one quantifier too wide.

The claim is now scoped to what actually holds: **no two tags held by processes alive at the same time collide.** That is the property a sweep needs, because a sweep can only ever signal a live process. The three components and the hole each one closes are spelled out — pid separates concurrent spawners, the per-process counter separates their panes, `CLOCK_MONOTONIC` separates a reused pid from its predecessor — and the residual is stated rather than hidden: the clock restarts at boot and the counter restarts per process, so two tags minted on *different boots* can collide. That is inert, because no process outlives the boot it was tagged on, so a colliding tag can never name a live process and the reaper is gone with its boot too.

The unit test's scope is now described honestly as well: `minted_tags_are_unique` measures the within-process half over 1000 mints, and the doc says so instead of letting a 1000-mint loop stand in for a global claim.

The changelog fragment carried the same absolute and is narrowed the same way, including the boot-scoping sentence.

File: src/agent_pty.rs [hunk 4]
```diff
@@ -1129,6 +1158,26 @@ pub fn spawn(opts: SpawnOptions<'_>) -> Result<AgentPty, AgentPtyError> {
...
+    // Still ahead of the fallible `take_writer` / `try_clone_reader` steps below,
+    // so a spawn that fails halfway leaves the child bounded rather than loose.
+    //
+    // `ChildGuard` covers that same halfway failure by killing the child
+    // outright, and this is deliberately not a substitute for it: the guard runs
+    // only while THIS process lives, which is the assumption #861 is about.
+    #[cfg(unix)]
+    if let Some(pid) = child_pid {
+        crate::wrap::arm_child_group_backstop(pid as libc::pid_t, lifetime_tag);
+    }
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply src/agent_pty.rs:1179]: Not fixed, deliberately, and this one is a judgment call I want a maintainer to make rather than settle myself. I raised it in the PR description before you flagged it, so we agree the trigger fires.

**Why I think the test cannot observe this diff.** Rule 12 exists to catch a semantic break behind a stable wire. Tracing the three things the diff adds to the daemon's spawn path:

1. `LifetimeTag::for_child()` reads two environment variables and returns `None` unless the test-only cap parses;
2. one `cmd.env(...)` on the child's `CommandBuilder` — applied straight to `cmd`, deliberately **not** through `opts.env`, which is what `spawn_env` captures and `respawn_agent_for_pane` replays, so nothing is persisted and no `agent_records` field changes shape or meaning;
3. `arm_child_group_backstop`, which returns before its `fork` when the cap is unset.

`PROTOCOL_VERSION` is untouched, no handler changed, and no field's meaning shifted. So in the configuration the cross-version run exercises, the diff is inert.

**And I measured that rather than asserting it,** with a positive control so the negative result means something:

```
PRODUCTION (cap unset): tag_in_child_environ=0  reapers_added=0
TEST       (cap=15)   : tag_in_child_environ=1  reapers_added=1
```

**Why I did not run it anyway.** Rule 12 is unusually explicit that a mis-sequenced run produces a *false green* which is worse than not running one — it names three separate traps (no live agent under the old daemon, the 30-second idle window, an unisolated log) and records two prior runs lost to two of them. The correct sequence requires driving a previous-release TUI and then a branch TUI interactively. I cannot do that reliably in this environment, so my realistic choice was between a run I could not vouch for and an honest abstention. I chose the abstention and said so up front.

I am not asking for the rule to be waived in general. If you want the dance run before merge, say so and it should be run by someone who can drive the two TUIs properly — or, if this shape of change is going to recur, rule 12 may deserve an explicit "inert in production" carve-out with the inertness *measured*, which is a change to CLAUDE.md rather than something to decide inside this PR. Leaving this thread open for that reason.

**Comment by vfarcic** [comment]
CI is settled. **All four required checks are green** (`build`, `build-macos`, `build-windows`, `security`), plus `nix`, `devbox`, `docs-build`, `desktop-web` and `windows-cross-check`.

`e2e-deterministic` is red, and it is **#818, not this change** — confirmed rather than assumed:

```
Summary [437.269s] 2599 tests run: 2597 passed (2 slow, 1 leaky), 2 failed, 0 skipped
  FAIL  dot-agent-deck::e2e_orchestration_remit orchestration_remit_001_start_role_compaction_reasserts_remit
  FAIL  dot-agent-deck::e2e_orchestration_remit orchestration_remit_005_non_start_role_clear_reasserts_nothing
```

Both are in `tests/e2e_orchestration_remit.rs`, which is the file [#818](https://github.com/vfarcic/dot-agent-deck/issues/818) names as its first of two causes — a 10 s timeout too tight for a runner, red on `main` and every branch since #789 merged. That lane is not a required check, and PR #837 fixes the remit half. Nothing in this diff touches orchestration remit, and the other 2597 tests in lane 1 pass.

Worth noting what that green lane-1 majority buys here, since it is more than a formality: those 2597 tests include the whole e2e tier's spawn surface driven through the real deck binary over PTYs — which is where `run_wrap_pty`'s half of the tag injection lives. My local tests exercise the pipe path (`run_wrap_pipe`), so CI is what covers the PTY one.

File: src/agent_pty.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (vfarcic) [reply src/agent_pty.rs:1179]: **Now fixed rather than deferred: I ran rule 12's cross-version manual test, and it passes.** You were right that the abstention left the requirement unmet, and that a maintainer should not have to adjudicate a judgment call that could simply be resolved by doing the run. So here it is, with the evidence rather than the assertion.

## Setup

Old side is the **real previous-release binary**, not a build-id override — an override runs the same code on both ends and so cannot observe a semantic break behind a stable wire, which is the entire point of this gate. `dot-agent-deck-linux-amd64` from the **v0.39.2** GitHub release. v0.39.2 deliberately, not the newer v0.39.3: `git merge-base --is-ancestor` says v0.39.2 **is** an ancestor of this branch and v0.39.3 is **not**, so the skew is unidirectional (branch strictly newer) instead of bidirectional.

New side is this branch at `0e40f70f`, `cargo build --locked --bin dot-agent-deck`.

Every process in the run shared one isolated environment, all five of the things rule 12 names: `DOT_AGENT_DECK_SOCKET`, `DOT_AGENT_DECK_ATTACH_SOCKET`, `HOME` and `DOT_AGENT_DECK_STATE_DIR` in a disk-backed sandbox; `DOT_AGENT_DECK_LOG` pointed into it, because the log path resolves separately and an otherwise-isolated daemon still appends into the operator's real `deck.log`; `DOT_AGENT_DECK_EXPERIMENTAL=0` pinned, because project-config discovery walks up from cwd and not from `$HOME` (the daemon log confirms it read the sandbox's own config: `experimental flag: OFF (from …/dad-xver-871/proj/.dot-agent-deck.toml)`); and `DOT_AGENT_DECK_IDLE_SHUTDOWN_SECS=0` against the 30-second window.

Sequence exactly as the rule prescribes: previous-release daemon → **previous-release TUI** opens the orchestration tab (the `Ctrl+N`, `Space`, `Right`, `Enter`, `Enter` sequence `e2e_orchestration_route_isolation.rs::open_orchestration_tab` uses) → that TUI killed, daemon and its agents left running → **branch TUI** → mismatch prompt → **declined**. Both TUIs were driven as real binaries in a PTY with the screen parsed by a terminal emulator, the same shape as `TuiDeck`.

## Confirmation the scenario was actually reached

This is the half that matters, since both of the rule's traps produce a *false green* that looks like a pass:

```
⚠  Daemon version mismatch  (2 agent(s) running)
   running daemon:  0.39.2-g364a182
   this binary:     0.39.2-g0e40f70f
   Restarting to upgrade will stop these agents:
   orchestrator
   coder
   [S] restart daemon and continue   [any other key] keep current daemon
```

- **`Attach protocol listening` lines in the isolated log: 1.** The tell for both traps — two would mean the daemon was replaced and the run had silently become same-version.
- Same daemon pid alive end to end, `readlink /proc/<pid>/exe` = the **v0.39.2** binary before *and* after the branch TUI attached; exactly **one** `daemon serve` process in the sandbox.
- The prompt named **both** live agents, so the no-live-agent trap was not taken, and it was declined with `n`.

## Results — the core flows still work

| flow | result |
| --- | --- |
| **delegate routes** | `delegate` (branch binary, from the orchestrator pane) → **rc 0**, no stderr. Old daemon: `Received delegate signal pane_id=1 targets=["coder"]`. Task pointer written into the worker pane's PTY (`Read .dot-agent-deck/worker-task-coder.md for your task.`) and the task text — carrying a unique sentinel — written to that file by the **old** daemon. |
| **work-done hook arrives** | `work-done` (branch binary, from the worker pane) → **rc 0**. Old daemon: `work-done: retired the outstanding delegation and cancelled its idle watch pane_id=2 role=coder` — i.e. matched to the commission the delegate armed, not treated as unsolicited. `work-done-coder.md` written with the sentinel, and the orchestrator pane received `Worker coder has completed their task.` So the **whole delegate → work-done round trip closed across the version boundary**, not just each leg in isolation. |
| **status hook** | `agent-event --type running` → rc 0. Run **last**, per the rule's closing note that doing it earlier makes the daemon classify the pane as `Pi` and can make a delivered delegate look undelivered. |
| agents survive | 2 still registered after the whole run. |

The one thing in the log that is *not* a defect: a `#249` "delegated worker went quiet" report. My role commands are deterministic sink scripts, not real agents, so they emit no agent event inside the 30 s window and the daemon correctly reports that to the orchestrator. It also quotes the worker pane's text back, which independently re-proves the pointer was delivered.

**Conclusion: no contract break.** `PROTOCOL_VERSION` unchanged is correct and no `.breaking.md` fragment applies. The measured inertness in the PR description now has the behavioural run behind it rather than standing alone, and the PR description has been corrected — it still claimed the test was not run.

## What the run does not prove

Stated rather than glossed. The roles are **stand-ins**, not real agents: this gate is about the TUI↔daemon contract, so I chose deterministic, credential-free delivery I could observe on disk over a real agent whose phrasing varies — but a break that only a real agent's traffic would trip is outside it. One previous release, one platform (Linux), one orchestration with one worker.

## One trap worth adding to the rule, because it cost the first attempt

`start = true` on a **worker** role makes every delegate to that role resolve to nothing, silently. `register_orchestration_role` reads that flag as `is_start_role` and inserts the pane into `orchestrator_pane_ids`, and `delegate_targets` filters exactly those panes out as targets. My first fixture set it on both roles; the delegate then failed with *"reached no worker for role(s): coder"* and the daemon logged `delegate: no worker pane found for role`. That reads precisely like the contract break this gate exists to find, and it was my fixture — `tests/fixtures/orchestration-route/.dot-agent-deck.toml` gets it right, with `start = true` on the orchestrator alone.

Worth noting that this is a trap of the same shape as the two the rule already documents, and in the *opposite* direction: rule 12's existing traps make a broken thing look green, this one makes a working thing look broken. I have not added it to CLAUDE.md here, because that is a change about rule 12 rather than about #861 and belongs in its own PR — happy to file it as a follow-up issue if you'd like it captured.

[the remark this answers is not in the document]
Comment (vfarcic) [reply src/agent_pty.rs:1179]: **Correction to my last comment: withdraw the "one trap worth adding to the rule" paragraph. There is no gap, and no follow-up issue should be filed.**

I checked before proposing the change and the premise does not survive it. `start = true` on two roles is already:

- **documented** — `docs/orchestration.md`'s configuration reference states it outright: *"`true` marks this role as the orchestrator. Exactly one role per orchestration must have `start = true`"*, and the `roles` row above it repeats the requirement;
- **validated as an error** — `src/config_validation.rs:352` counts the start roles and pushes a `Severity::Error` when the count is not exactly 1;
- **unit-tested** — `config_validation::tests::orchestration_multiple_start_roles_is_error`;
- **reported by a shipped command**, which I confirmed by running it rather than by reading the code. My original broken fixture:

```
$ dot-agent-deck validate --path <fixture>
[error] 'xver': orchestration must have exactly one role with start = true
exit=1
```

and the corrected one exits 0.

So what actually happened is operator error, not a footgun: I hand-edited a config into a state the project documents as invalid and rejects with a non-zero exit, and I never ran `validate` on it. Describing that as a trap of the same shape as rule 12's sequencing traps was wrong — those are cases where the tooling cannot tell you, and this is a case where it does, loudly.

Adding it to CLAUDE.md would have been worse than redundant: it would assert that an enforced validation error is an undetected hazard, which is exactly the over-wide claim rule 17 is about. The cross-version run itself and everything else in the comment above stand unchanged — the failure was caught, diagnosed and corrected before the result was reported, and the reported run is the corrected one.

**Review by prageethw — APPROVED** [review]
Reviewed the tag mechanism (src/lifetime_tag.rs), the agent_pty::spawn/wrap.rs wiring, and the new tests. The fix is well-scoped: gated entirely on `DOT_AGENT_DECK_TEST_MAX_LIFETIME_SECS` so production mints/exports/sweeps nothing, the needle-matching bug and cleanup-can-kill-strangers findings from Greptile are both fixed (exact NUL-delimited entry match, single-pid SIGKILL instead of group kill), and the tag-reuse-per-subtree design avoids the single-point-of-failure the issue is about. Tests cover the escapee case with SIGKILL'd wrapper, the tag-carries-through-execve property, and the substring/pid-parsing edge cases. CI is green on build/build-macos/build-windows/security; e2e-deterministic's failure is the confirmed pre-existing orchestration_remit race (fixed by #837 in this same queue), unrelated to this diff. Rule 12 cross-version test was run and documented in the PR body. All four review threads are resolved. LGTM.
